flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceStreamTask.java

```diff
@@ -265,6 +265,14 @@ protected void declineCheckpoint(long checkpointId) {
 
         @Override
         public void run() {
+            if (operatorChain.isFinishedOnRestore()) {
```

Comment:
I am fine with that solution, but I thought why can't we do it a bit earlier and don't even start the thread? I think we could move this block to the `processInput` method (plus a small adjustment to the `interruptSourceThread` to skip it if the operator chain has finished. WDYT?

Reply:
I also think would be better to directly skipping the startup of source thread. With this change the callback of `CompletionFuture` would be executed in the main thread, but it should also be correct. I changed it to not start the thread~